Flow.Launcher/ViewModel/MainViewModel.cs

```diff
@@ -1440,26 +1447,35 @@ public void UpdateResultView(ICollection<ResultsForUpdate> resultsForUpdates)
             }
 #endif
 
-            foreach (var metaResults in resultsForUpdates)
+            try
```

Comment:
could you explain for the try?

Reply:
It could happen when calling `ResultsUpdated.Invoke` event in` IResultUpdate` interface so fast that main view model cannot complete update event (still execute foreach sentence).

Take `WebSearch` plugin for example,

```
public async Task<List<Result>> QueryAsync(Query query, CancellationToken token)
{
    var results = new List<Result>();

    foreach (...)
    {
        ...

        results.Add(result);  // edit this List

        ...

        // Here you use this List to give to MainViewModel for updating
        // If you do the time-consuming work after this so fast, you will edit this list when updating the MainViewModel
        // So we need to use Try to avoid this problem
        ResultsUpdated?.Invoke(this, new ResultUpdatedEventArgs
        {
            Results = results,
            Query = query
        });

        ... // do time-consuming work
    }

    return results;
}
```

Reply:
Make sense, but that also means we need additional synchronization? I remember the ConcurrentDictionary is already providing a enumerator that is thread safe?

Reply:
How about a `ConcurrentBag`?

Reply:
The key of the problem is that this plugin has not tell the developer to do thread-safe operation in `QueryAsync` (like return a copy of the list as the `Results` in `ResultUpdatedEventArgs`). So the developer will possibly return the same `List` to the `ResultUpdatedEventArgs` and the result of the `QueryAsync` function.

If you want to fully address this problem, you need to add additional synchronization.

For example, you can change the definition of the `ResultUpdatedEventArgs` to
```
public class ResultUpdatedEventArgs : EventArgs
{
    public ConcurrentBag<Result> Results;  // thread-safe
    public Query Query;
    public CancellationToken Token { get; init; }
}
```.

Or, you need to ask the developer to return a copy of the `List` so that the plugin will not change the `List` for updating  the view.

For example,
```
public async Task<List<Result>> QueryAsync(Query query, CancellationToken token)
{
    var results = new List<Result>();

    foreach (...)
    {
        ...

        results.Add(result);

        ...

        var resultsCopy = results.Clone();  // create a copy of the results
        ResultsUpdated?.Invoke(this, new ResultUpdatedEventArgs
        {
            Results = resultsCopy,
            Query = query
        });

        ...
    }

    return results;
}
```

Anyway, I think this is the most developer-friendly solution for this problem and developers don not need to redevelop their plugins to fix this.

Reply:
@taooceros What do you mean by `ConcurrentDictionary` and `ConcurrentBag`?

Reply:
Hold on, I remember this is a producer/consumer model with appropriate synchronization through the `Channel` class. Let me recheck.

Reply:
I am a little bit suspicious of this change. Is it possible for you to revert just this part so we can merge first. Then Include it in a new pr so we can discuss specifically about the potential synchronization problem.

Reply:
@taooceros Sure! Please check!